src/color.rs

```diff
@@ -160,3 +165,67 @@ impl<C: Color> fmt::Display for Bg<C> {
         self.0.write_bg(f)
     }
 }
+
+/// Types that allow detection of the colors they support
+pub trait DetectColors {
+    /// How many ANSI colors are supported (from 8 to 256)?
+    ///
+    /// Beware: the information given isn't authoritative, it's infered through escape codes or the
+    /// value of `TERM`, more colors may be available.
+    fn available_colors(&mut self) -> io::Result<u16>;
+}
+
+impl<W: Write> DetectColors for RawTerminal<W> {
+    fn available_colors(&mut self) -> io::Result<u16> {
+        let mut stdin = async_stdin();
+
+        if detect_color(self, &mut stdin, 0)? {
+            // OSC 4 is supported, detect how many colors there are
+            let mut min = 8;
+            let mut max = 256;
+            let mut i;
+            while min + 1 < max {
```

Comment:
While it is somewhat obvious what happens here, it could need a small explanatory comment.